src/main/java/org/broadinstitute/hellbender/utils/downsampling/ReservoirDownsampler.java

```diff
@@ -0,0 +1,181 @@
+package org.broadinstitute.hellbender.utils.downsampling;
+
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ * Reservoir Downsampler: Selects n reads out of a stream whose size is not known in advance, with
+ * every read in the stream having an equal chance of being selected for inclusion.
+ *
+ * An implementation of "Algorithm R" from the paper "Random Sampling with a Reservoir" (Jeffrey Scott Vitter, 1985)
+ *
+ * @author David Roazen
+ */
+public final class ReservoirDownsampler<T extends GATKRead> extends ReadsDownsampler<T> {
+
+    /**
+     * size of our reservoir -- ie., the maximum number of reads from the stream that will be retained
+     * (not including any undiscardable items)
+     */
+    private final int targetSampleSize;
+
+    /**
+     * if true, this downsampler will be optimized for the case
+     * where most of the time we won't fill up anything like the
+     * targetSampleSize elements.  If this is false, we will allocate
+     * internal buffers to targetSampleSize initially, which minimizes
+     * the cost of allocation if we often use targetSampleSize or more
+     * elements.
+     */
+    private final boolean expectFewOverflows;
+
+    /**
+     * At times this can be a linked list or an array list, depending on how we're accessing the
+     * data and whether or not we're expecting few overflows
+     */
+    private List<T> reservoir;
+
+    /**
+     * Are we currently using a linked list for the reservoir?
+     */
+    private boolean isLinkedList;
+
+    /**
+     * Count of the number of reads seen that were actually eligible for discarding. Used by the reservoir downsampling
+     * algorithm to ensure that all discardable reads have an equal chance of making it into the reservoir.
+     */
+    private int totalDiscardableReadsSeen;
```

Comment:
Rename to `totalReadsSeen` and update the comment, since the concept of "undiscardable" reads is gone.